.changeset/wet-taxis-share.md

```diff
@@ -0,0 +1,5 @@
+---
+"miniflare": patch
+---
+
+fix: make the magic proxy able to serialize objects containing functions
```

Comment:
Can you make clear that this fix doesn't allow those functions to be callable

Reply:
Is that true? JS RPC does allow methods and callback to be called I believe.

Reply:
This isn't JS RPC – it's the miniflare magic proxy

Reply:
But isn't it using JS RPC under the hood? Perhaps I am misunderstanding.

Reply:
@petebacondarwin no this is unrelated to JS RPC 🙂, it is using the magic proxy and saving such object on the heap (in workerd) so that they can be referenced later.

JS RPC is implemented in the same sort of way: https://github.com/cloudflare/workers-sdk/blob/66bdad08834b403100d1e4d6cd507978cc50eaba/packages/miniflare/src/workers/core/proxy.worker.ts#L55-L57
but one is not necessary related to the other

Reply:
PS: I've updated the changeset as I can see that `serialize` here is not really the right word and can lead to confusion 🙂 
(my changes are not actually serializing such objects but allowing us to handle them)

Reply:
Ah yeah sorry I didn't see any implementation changes so assumed you were simply allowing the object to be _serialised_ whereas before it was prevented from _proxying_.

Am I right to now think that proxying objects with functions _always_ worked? If so, do we know why we explicitly prevented it?

Reply:
> Am I right to now think that proxying objects with functions always worked?

No, they never worked, only the rpc stubs ones but that's because we're specifically handling them ad-hoc

>  If so, do we know why we explicitly prevented it?

The thing is that with my changes we check if an object contains functions and if so we mark as a non-plain object.
Basically my changes make the `!isPlainObject(object)` trigger here:
https://github.com/cloudflare/workers-sdk/blob/c0b057b9d196dc46f54530f91e8c87dee88d6cfa/packages/miniflare/src/workers/core/proxy.worker.ts#L88-L93
making it so that the object doesn't get sent to devalue which would otherwise complain when trying to serialize it (saying that it doesn't know how to serialize functions).